File: webmanager/handlerevents/befdownloadfile/befdownloadfile.js
// Обработчик событий (сохранение лога действия плюс необходимые проверки перед загрузкой)
function befdownloadfile(reload = false, rdata= null) {
	if (g_downloading) return;                    // Идет загрузка файла
    function padStr(i) {
        return (i < 10) ? "0" + i : "" + i;
    }
	if (!reload) {
	    // Проверка корректности введенных данных ========
	    if (g_sendfile == null) {
		   alert('Необходимо выбрать файл');
		   return;
	    };
	    var date = $("input#date").datepicker('getDate')   // Выбранная дата в календаре
	        if (date == null) {
		    alert('Необходимо выбрать дату');
		    return;
	    }	
        var str_date = padStr(date.getFullYear()) + '_' + padStr(1 + date.getMonth()) + '_' + padStr(date.getDate());
        var reason = $('#reasons').val();
	    if (reason == '') {
		    alert('Необходимо ввести причину');
		    return;
	    }
        var user = $("#combo_users option:selected" ).text();
		document.cookie = "user=" + user;                  // Сохраняем пользователя в cookie
		var descr = $('#descr').prop('value');
	    var unzip = $("#enunzip").is(':checked');
	    if (unzip) {
		    if (g_sendfile.name.slice(-3) != 'zip') {
		        alert('В хранилище можно распаковать только ".zip" файл');
		        return;			
		    }
			var res = confirm('Действительно распаковать файлы в хранилище?');
			if (!res) return;
	    }
	    var replace = false;
	}
	// ===============================================
	if (!reload)
	    var querydata = {'folder': g_selfolder, 'filename': g_sendfile.name, 'sdate': str_date, 'reason': reason, 
	                     'user': user, 'descr': descr, 'unzip': unzip, 'replace': replace};
    else var querydata = rdata;
	// Передаем серверу имя файла и имя папки для него
    $.getJSON('http://' + g_rhost + '/befdownloadfile' + '?path=' + g_fullpath + '&data=' + JSON.stringify(querydata),
        function(data) {
			if (data == 'notfound') {   // Директория не найдена, кто-то удалил ее пока браузер был открыт
			    alert('Выделенная в дереве папка не найдена. Попробуйте обновить дерево!');
				return;
			}
			if (data == 'notexists') {      // Папка не найдена, кто-то удалил эту папку пока браузер был открыт
			    alert('Папка для загрузки файла не найдена. Попробуйте обновить дерево!')
				return;
			}
			if (data == 'collision') {      // Попытка записать уже записываемый файл
			    alert('Данный файл уже загружается с другого хоста.\nПопробуйте загрузить файл через некоторое время!')
				return;
			}
			if (data == 'existsfolder') {   // Папка куда будут распакованы файлы существует
                var res = confirm('Папка, куда распаковываются файлы уже существует. Перезаписать?');
	            if (res == true) {
					querydata['replace'] = true;
				    befdownloadfile(true, querydata);    // Вызываем себя же
				}
				return;
			}
			if (data == 'existsfile') {     // Файл с таким именем уже существует
                var res = confirm('Файл с таким именем уже существует. Перезаписать?');
	            if (res == true) {
					querydata['replace'] = true;
				    befdownloadfile(true, querydata);    // Вызываем себя же
				}
				return;
			}
			if (data == 'success') {         // Имя файла передано серверу
			    downloadfile();              // Загрузка файла на сервер (обработчик в handlerevents/downloadfile)
			}
		}
    )			
}

File: webmanager/handlerevents/befdownloadfile/befdownloadfile.py
import aiopg
import os
import json

# Класс обработчик событий (выполнение действий до загрузки файла)
class BefDownloadFile:
    # Конструктор   
    def __init__(self, app, web):
        app.add_routes([web.get('/befdownloadfile', self.befdownloadfile)])     # Добавление маршрута

    # Обработчик события до загрузки файла
    async def befdownloadfile(self, request):
        params = request.rel_url.query
        fullpath = self.get_fullpath(params['path'], request.remote)            # Полный путь к выделенной папке
        if not os.path.exists(fullpath):                                        # Нет выделенной в дереве папки
            return self.web.json_response('notfound')

        data = json.loads(params['data']);
        folder = data['folder']; filename = data['filename']; sdate = data['sdate']
        reason = data['reason']; user = data['user']; descr = data['descr']
        unzip = data['unzip']; replace = data['replace']

        try:
            async with aiopg.create_pool(self.dsn) as pool:
                async with pool.acquire() as conn:
                    async with conn.cursor() as cur:
                        await cur.execute('SELECT "REASONS" FROM public."SETTINGS"')
                        async for row in cur:
                            reasons = row[0]                                       # Возможные причины добавления данных
                        if reasons == None: reasons = []
                        if not (reason in reasons):                                # Если такой причины добавления нет добавляем ее
                            reasons.append(reason)
                            await cur.execute('UPDATE public."SETTINGS" SET "REASONS" = '"'{}{}{}'"''.format('{', ','.join(reasons), '}'))
        except Exception:
            print('Ошибка при обновлении причины загрузки')
                        
        if not os.path.exists(fullpath + '/' + folder):                         # Папка, в которую загружается файл не найдена
            return self.web.json_response('notexists')
        
        self.downdata[request.remote] = (fullpath + '/' + folder, filename, sdate, reason, user, descr, unzip)
        
        downname = fullpath + '/' + folder + '/' + filename                     # Имя загружаемого файла
        if downname in self.downfiles:                                          # Файл уже загружается с другого хоста
            return self.web.json_response('collision')

        if replace:
            self.downfiles.append(downname)                                     # Сохраняем имя загружаемого файла
            self.downhosts[downname] = request.host
            return self.web.json_response('success')
        
        path = fullpath + '/' + folder
        if unzip:                                 # Распаковать файл
            folder = sdate #folder = sdate.replace('_', '')
            if os.path.exists(path + '/' + folder):
                return self.web.json_response('existsfolder')
        else:
            fullname = path + '/' + sdate + '_' + filename
            if os.path.exists(fullname):
                return self.web.json_response('existsfile')
        
        self.downfiles.append(downname)                                        # Сохраняем имя загружаемого файла
        self.downhosts[downname] = request.host
        return self.web.json_response('success')
        
